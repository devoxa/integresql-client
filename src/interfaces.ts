export interface IntegreSQLDatabaseConfig {
  username: string
  password: string
  host: string
  port: number
  database: string
}

export interface InitializeTemplateResponse {
  database: {
    templateHash: string
    config: IntegreSQLDatabaseConfig
  }
}

export interface GetTestDatabaseResponse {
  id: number
  database: {
    templateHash: string
    config: IntegreSQLDatabaseConfig
  }
}
