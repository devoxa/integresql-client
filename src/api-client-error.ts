export class IntegreSQLApiClientError extends Error {
  responseStatus: number
  responseText: string
}

export interface CreateIntegreSQLApiClientErrorOptions {
  message: string
  responseStatus: number
  responseText: string
}

export function createIntegreSQLApiClientError(options: CreateIntegreSQLApiClientErrorOptions) {
  const error = new IntegreSQLApiClientError(options.message)
  error.responseStatus = options.responseStatus
  error.responseText = options.responseText
  return error
}
