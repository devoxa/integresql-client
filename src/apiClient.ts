import { createIntegreSQLApiClientError } from './apiClientError'
import { GetTestDatabaseResponse, InitializeTemplateResponse } from './interfaces'

export interface IntegreSQLApiClientOptions {
  url: string
}

export class IntegreSQLApiClient {
  baseUrl: string

  constructor(options: IntegreSQLApiClientOptions) {
    const url = new URL(options.url)
    this.baseUrl = url.toString() + 'api/v1'
  }

  async request<T>(
    method: 'GET' | 'POST' | 'PUT' | 'DELETE',
    endpoint: string,
    body?: Record<string, unknown>
  ): Promise<T> {
    const url = this.baseUrl + endpoint

    const response = await fetch(url, {
      method,
      body: body ? JSON.stringify(body) : undefined,
      headers: { 'Content-Type': 'application/json' },
    })

    const text = await response.text()

    if (response.status >= 400) {
      throw createIntegreSQLApiClientError({ responseStatus: response.status, responseText: text })
    }

    return text ? (JSON.parse(text) as T) : (null as T)
  }

  async initializeTemplate(hash: string): Promise<InitializeTemplateResponse> {
    return this.request<InitializeTemplateResponse>('POST', '/templates', { hash })
  }

  async finalizeTemplate(hash: string): Promise<null> {
    return this.request<null>('PUT', `/templates/${hash}`)
  }

  async discardTemplate(hash: string): Promise<null> {
    return this.request<null>('DELETE', `/templates/${hash}`)
  }

  async discardAllTemplates(): Promise<null> {
    return this.request<null>('DELETE', '/admin/templates')
  }

  async getTestDatabase(hash: string): Promise<GetTestDatabaseResponse> {
    return this.request<GetTestDatabaseResponse>('GET', `/templates/${hash}/tests`)
  }

  async reuseTestDatabase(hash: string, id: number): Promise<null> {
    return this.request<null>('DELETE', `/templates/${hash}/tests/${id}`)
  }

  async recreateTestDatabase(hash: string, id: number): Promise<null> {
    return this.request<null>('POST', `/templates/${hash}/tests/${id}/recreate`)
  }
}
