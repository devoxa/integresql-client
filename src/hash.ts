import crypto from 'crypto'
import glob from 'fast-glob'
import fs from 'fs'
import path from 'path'

export async function sha1HashFiles(globPatterns: Array<string>): Promise<string> {
  const files = await glob(globPatterns)
  const filePaths = files.map((file) => path.join(process.cwd(), file))

  const fileHashes = await Promise.all(filePaths.map(sha1HashFile))
  return sha1HashString(fileHashes.join(''))
}

export function sha1HashString(value: string): string {
  return crypto.createHash('sha1').update(value).digest('hex')
}

export function sha1HashFile(filePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const hash = crypto.createHash('sha1')
    const readStream = fs.createReadStream(filePath)

    // istanbul ignore next
    readStream.on('error', (err) => reject(err))

    readStream.on('data', (chunk) => hash.update(chunk))
    readStream.on('end', () => resolve(hash.digest('hex')))
  })
}
