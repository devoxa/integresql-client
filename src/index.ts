export * from './api-client'
export * from './api-client-error'
export * from './client'
export * from './interfaces'
