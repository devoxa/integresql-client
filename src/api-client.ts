import fetch from 'node-fetch'
import path from 'path'
import { createIntegreSQLApiClientError } from './api-client-error'
import { GetTestDatabaseResponse, InitializeTemplateResponse } from './interfaces'

export interface IntegreSQLApiClientOptions {
  url: string
}

export class IntegreSQLApiClient {
  baseUrl: string

  constructor(options: IntegreSQLApiClientOptions) {
    const url = new URL(options.url)
    this.baseUrl = url.toString() + 'api/v1'
  }

  async request<T>(
    method: 'GET' | 'POST' | 'PUT' | 'DELETE',
    endpoint: string,
    body?: Record<string, unknown>
  ): Promise<T> {
    const url = path.join(this.baseUrl, endpoint)

    const response = await fetch(url, {
      method,
      body: body ? JSON.stringify(body) : undefined,
      headers: { 'Content-Type': 'application/json' },
    })

    const text = await response.text()

    if (response.status >= 400) {
      throw createIntegreSQLApiClientError({ responseStatus: response.status, responseText: text })
    }

    return text ? JSON.parse(text) : null
  }

  async initializeTemplate(hash: string) {
    return this.request<InitializeTemplateResponse>('POST', '/templates', { hash })
  }

  async finalizeTemplate(hash: string) {
    return this.request<null>('PUT', `/templates/${hash}`)
  }

  async discardTemplate(hash: string) {
    return this.request<null>('DELETE', `/templates/${hash}`)
  }

  async discardAllTemplates() {
    return this.request<null>('DELETE', `/admin/templates`)
  }

  async getTestDatabase(hash: string) {
    return this.request<GetTestDatabaseResponse>('GET', `/templates/${hash}/tests`)
  }

  async reuseTestDatabase(hash: string, id: number) {
    return this.request<null>('DELETE', `/templates/${hash}/tests/${id}`)
  }
}
