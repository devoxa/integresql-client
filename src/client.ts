import { IntegreSQLApiClient } from './api-client'
import { IntegreSQLApiClientError } from './api-client-error'
import { sha1HashFiles } from './hash'
import { IntegreSQLDatabaseConfig } from './interfaces'

export interface IntegreSQLClientOptions {
  url: string
}

export class IntegreSQLClient {
  api: IntegreSQLApiClient

  constructor(options: IntegreSQLClientOptions) {
    this.api = new IntegreSQLApiClient({ url: options.url })
  }

  async initializeTemplate(
    hash: string,
    setupCallback: (databaseConfig: IntegreSQLDatabaseConfig) => Promise<void>
  ): Promise<void> {
    // Attempt to create a new template database identified through the hash
    let databaseConfig
    try {
      const response = await this.api.initializeTemplate(hash)
      databaseConfig = response.database.config
    } catch (err) {
      // Some other process has already created a template database for this hash,
      // so we can just consider it ready at this point.
      if (err instanceof IntegreSQLApiClientError && err.responseStatus === 423) {
        return
      }

      throw err
    }

    // The template database is created and ready for setup (apply migrations / seed fixtures)
    try {
      await setupCallback(databaseConfig)
    } catch (err) {
      // If we encountered any errors during setup we discard the template database
      await this.api.discardTemplate(hash)

      throw err
    }

    // The template database is setup and ready for tests
    await this.api.finalizeTemplate(hash)

    // Warm up the template database, since getting the first test database can take a second or two
    await this.getTestDatabase(hash)
  }

  async getTestDatabase(hash: string): Promise<IntegreSQLDatabaseConfig> {
    const response = await this.api.getTestDatabase(hash)
    return response.database.config
  }

  // --- Helpers --- //

  async hashFiles(globPatterns: Array<string>): Promise<string> {
    return sha1HashFiles(globPatterns)
  }

  databaseConfigToConnectionUrl(databaseConfig: IntegreSQLDatabaseConfig): string {
    const { username, password, host, port, database } = databaseConfig
    return `postgresql://${username}:${password}@${host}:${port}/${database}`
  }
}
