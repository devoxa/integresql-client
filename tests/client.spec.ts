import { IntegreSQLApiClient } from '../src/apiClient'
import { createIntegreSQLApiClientError } from '../src/apiClientError'
import { IntegreSQLClient } from '../src/client'
import { sha1HashFiles } from '../src/hash'
import { IntegreSQLDatabaseConfig } from '../src/interfaces'

jest.mock('../src/apiClient')
const mockIntegreSQLApiClient = jest.mocked(IntegreSQLApiClient)
const MockIntegreSQLApiClient = mockIntegreSQLApiClient.prototype

jest.mock('../src/hash')
const mockSha1HashFiles = jest.mocked(sha1HashFiles)

const client = new IntegreSQLClient({ url: 'http://localhost:5000' })

const mockDatabaseConfig: IntegreSQLDatabaseConfig = {
  username: 'user',
  password: 'pass',
  host: 'localhost',
  port: 5432,
  database: 'mock_database_0',
}

describe('client', () => {
  beforeEach(() => {
    for (const key of Object.keys(MockIntegreSQLApiClient)) {
      const value = MockIntegreSQLApiClient[key as keyof typeof MockIntegreSQLApiClient]

      if (typeof value === 'function') {
        value.mockReset()
      }
    }
  })

  test('instantiates the API client', () => {
    expect(client.api).toBeInstanceOf(IntegreSQLApiClient)
    expect(typeof client.api.initializeTemplate).toEqual('function')
  })

  test('can initialize a template', async () => {
    const mockSetupCallback = jest.fn().mockImplementation(async () => 'noop')

    MockIntegreSQLApiClient.initializeTemplate.mockImplementation(async () => ({
      database: {
        templateHash: 'mock-hash',
        config: { ...mockDatabaseConfig, database: 'mock_template_1' },
      },
    }))

    MockIntegreSQLApiClient.getTestDatabase.mockImplementation(async () => ({
      id: 1,
      database: {
        templateHash: 'mock-hash',
        config: { ...mockDatabaseConfig, database: 'mock_database_1' },
      },
    }))

    await client.initializeTemplate('mock-hash', mockSetupCallback)

    expect(MockIntegreSQLApiClient.initializeTemplate.mock.calls).toEqual([['mock-hash']])
    expect(mockSetupCallback.mock.calls).toMatchSnapshot()
    expect(MockIntegreSQLApiClient.discardTemplate.mock.calls).toEqual([])
    expect(MockIntegreSQLApiClient.finalizeTemplate.mock.calls).toEqual([['mock-hash']])
    expect(MockIntegreSQLApiClient.getTestDatabase.mock.calls).toEqual([['mock-hash']])
  })

  test('can skip initializing a template when another process is already doing it', async () => {
    const mockSetupCallback = jest.fn()

    MockIntegreSQLApiClient.initializeTemplate.mockImplementation(async () => {
      const responseStatus = 423
      const responseText = JSON.stringify({ message: 'template is already initialized' })
      throw createIntegreSQLApiClientError({ responseStatus, responseText })
    })

    await client.initializeTemplate('mock-hash', mockSetupCallback)

    expect(MockIntegreSQLApiClient.initializeTemplate.mock.calls).toEqual([['mock-hash']])
    expect(mockSetupCallback.mock.calls).toEqual([])
    expect(MockIntegreSQLApiClient.discardTemplate.mock.calls).toEqual([])
    expect(MockIntegreSQLApiClient.finalizeTemplate.mock.calls).toEqual([])
    expect(MockIntegreSQLApiClient.getTestDatabase.mock.calls).toEqual([])
  })

  test('can rethrow error when initializing a template', async () => {
    const mockSetupCallback = jest.fn()

    MockIntegreSQLApiClient.initializeTemplate.mockImplementation(async () => {
      const responseStatus = 500
      const responseText = JSON.stringify({ message: 'connect: connection refused' })
      throw createIntegreSQLApiClientError({ responseStatus, responseText })
    })

    await expect(
      client.initializeTemplate('mock-hash', mockSetupCallback)
    ).rejects.toThrowErrorMatchingSnapshot()

    expect(MockIntegreSQLApiClient.initializeTemplate.mock.calls).toEqual([['mock-hash']])
    expect(mockSetupCallback.mock.calls).toEqual([])
    expect(MockIntegreSQLApiClient.discardTemplate.mock.calls).toEqual([])
    expect(MockIntegreSQLApiClient.finalizeTemplate.mock.calls).toEqual([])
    expect(MockIntegreSQLApiClient.getTestDatabase.mock.calls).toEqual([])
  })

  test('can discard a template when the setup callback fails', async () => {
    const mockSetupCallback = jest.fn().mockImplementation(async () => {
      throw new Error('Some migration error')
    })

    MockIntegreSQLApiClient.initializeTemplate.mockImplementation(async () => ({
      database: {
        templateHash: 'mock-hash',
        config: { ...mockDatabaseConfig, database: 'mock_template_1' },
      },
    }))

    await expect(client.initializeTemplate('mock-hash', mockSetupCallback)).rejects.toThrowError(
      'Some migration error'
    )

    expect(MockIntegreSQLApiClient.initializeTemplate.mock.calls).toEqual([['mock-hash']])
    expect(mockSetupCallback.mock.calls).toMatchSnapshot()
    expect(MockIntegreSQLApiClient.discardTemplate.mock.calls).toEqual([['mock-hash']])
    expect(MockIntegreSQLApiClient.finalizeTemplate.mock.calls).toEqual([])
    expect(MockIntegreSQLApiClient.getTestDatabase.mock.calls).toEqual([])
  })

  test('can get a test database', async () => {
    MockIntegreSQLApiClient.getTestDatabase.mockImplementation(async () => ({
      id: 2,
      database: {
        templateHash: 'mock-hash',
        config: { ...mockDatabaseConfig, database: 'mock_database_2' },
      },
    }))

    const databaseConfig = await client.getTestDatabase('mock-hash')

    expect(MockIntegreSQLApiClient.getTestDatabase.mock.calls).toEqual([['mock-hash']])
    expect(databaseConfig).toMatchSnapshot()
  })

  test('can call the hash files helper', async () => {
    await client.hashFiles(['./prisma/schema.prisma', './src/__fixtures__/**/*'])

    expect(mockSha1HashFiles.mock.calls).toMatchSnapshot()
  })

  test('can convert the database config into a connection url', () => {
    expect(client.databaseConfigToConnectionUrl(mockDatabaseConfig)).toMatchSnapshot()
  })
})
