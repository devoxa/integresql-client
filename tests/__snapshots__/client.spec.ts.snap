// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client can call the hash files helper 1`] = `
[
  [
    [
      "./prisma/schema.prisma",
      "./src/__fixtures__/**/*",
    ],
  ],
]
`;

exports[`client can convert the database config into a connection url 1`] = `"postgresql://user:pass@localhost:5432/mock_database_0"`;

exports[`client can discard a template when the setup callback fails 1`] = `
[
  [
    {
      "database": "mock_template_1",
      "host": "localhost",
      "password": "pass",
      "port": 5432,
      "username": "user",
    },
  ],
]
`;

exports[`client can get a test database 1`] = `
{
  "database": "mock_database_2",
  "host": "localhost",
  "password": "pass",
  "port": 5432,
  "username": "user",
}
`;

exports[`client can initialize a template 1`] = `
[
  [
    {
      "database": "mock_template_1",
      "host": "localhost",
      "password": "pass",
      "port": 5432,
      "username": "user",
    },
  ],
]
`;

exports[`client can rethrow error when initializing a template 1`] = `"API request to IntegreSQL failed: {"message":"connect: connection refused"} (Status 500)"`;
